### PR TITLE
[FLINK-22483][runtime][coordination] Recover checkpoints when JobMaster gains leadership

### PR DESCRIPTION
**Pull request by edu05**
## What is the purpose of the change

Moves the synchronization of checkpoint pointers with in-memory state out of the JobManager's (which would make it unresponsive to RPC for the duration of the sync)                                    

## Brief change log
 - Deleted old place where checkpoint pointer sync was taking placed
 - Added checkpoint pointer sync immediately after checkpoint store creation
 - Refactored test to be a bit more meaningful

## Verifying this change

 - Unit tests
 - Ideally the creator of the JIRA ticket would verify if the PR fixes their issue

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector:  no 

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by edu05** [comment]
@flinkbot run azure

**Comment by edu05** [comment]
@flinkbot run azure

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SchedulerUtilsTest.java [hunk 1]
```diff
@@ -21,17 +21,30 @@
...
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.tasks.JobCheckpointingSettings;
 import org.apache.flink.util.TestLogger;
 
 import org.junit.Test;
 
 import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
```
Comment (rmetzger) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SchedulerUtilsTest.java:35]: We try to not use Mocking in Flink (only in very rare cases, such as external systems not providing proper interfaces). See: https://flink.apache.org/contributing/code-style-and-quality-common.html#design-for-testability

**Comment by dmvk** [comment]
Hi @edu05, for CI failures, from a quick look, the flaky tests you're hitting have been fixed super recently (~ tuesday). Can you please try to rebase this on master?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SchedulerUtilsTest.java (hunk 1 above)
Comment (edu05) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SchedulerUtilsTest.java]: Having read that link and the presentation I can't say I agree with said decision but you're the boss!

**Comment by edu05** [comment]
@dmvk @rmetzger finally a passing build and I hope to have addressed all of the PR comments, please let me know. Happy Friday :)

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java [hunk 2]
```diff
@@ -1,213 +0,0 @@
...
-        private static volatile boolean failed = false;
-        private static volatile boolean recovered = false;
-
-        public static void reset() {
-            started = failed = recovered = false;
-        }
-
-        @Override
-        public void recover() throws Exception {
-            if (!started) {
```
Comment (dmvk) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java:148]: this needs to be updated with an extra "warmup" step

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java [hunk 3]
```diff
@@ -1,213 +0,0 @@
...
-import static org.apache.flink.api.common.restartstrategy.RestartStrategies.fixedDelayRestart;
-import static org.apache.flink.configuration.JobManagerOptions.SchedulerType.Adaptive;
-import static org.apache.flink.util.Preconditions.checkState;
-import static org.junit.Assume.assumeFalse;
-
-/**
- * Test that failure on recovery leads to job restart if configured, so that transient recovery
- * failures can are mitigated.
- */
-public class CheckpointStoreITCase extends TestLogger {
```
[the remark this answers is not in the document]
Comment (dmvk) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java:62]: Also we need an acceptance test case for the actual issue, ideally by constructing a DefaultCompletedCheckpointStore with blocking state handles and asserting that RPC remains responsive during recovery.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SchedulerUtilsTest.java [hunk 4]
```diff
@@ -40,16 +69,55 @@ public void testSettingMaxNumberOfCheckpointsToRetain() throws Exception {
...
-                completedCheckpointStore.getMaxNumberOfRetainedCheckpoints());
+        final CompletedCheckpointStore actualCompletedCheckpointStore =
+                SchedulerUtils.createCompletedCheckpointStoreIfCheckpointingIsEnabled(
+                        jobGraph, jobManagerConfig, classLoader, checkpointRecoveryFactory, log);
+
+        assertTrue(expectedCompletedCheckpointStore.recovered);
+        assertEquals(expectedCompletedCheckpointStore, actualCompletedCheckpointStore);
+    }
+
+    private class RecoveryRecordingCompletedCheckpointStore implements CompletedCheckpointStore {
```
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SchedulerUtilsTest.java:80]: This can be simplified by just extending StandalonCompletedCheckpointStore, then you just need to override a single method. Also this class can be static.

**Review by dmvk — CHANGES_REQUESTED** [review]
Hi @edu05, thanks for the contribution, this will be a great improvement! ;) I think you're on the right track, but there is still some work to be done, especially on the acceptance test side. I've added few comments with bit more context and suggestions.

**Comment by edu05** [comment]
@dmvk The idea was not to remove the test just because it failed, I should have left an explanation for its removal, apologies! The CheckpointStoreITCase effectively tested

1. Restore happens within CheckpointStore
2. If restore fails job is restarted

I thought the intention was to never recover() inside CheckpointStore and to fail entirely if the new invocation failed, for that reason I removed the test. I was left in doubt about the need for a new acceptance test, your suggestion of a DefaultCompletedCheckpointStore with blocking state handles is exactly that!

I'll implement all your suggestions and ping you back :)

**Comment by edu05** [comment]
Hi @dmvk I've addressed most of the comments, I'm just missing something to complete the new acceptance test. If you take a look at CheckpointStoreITCase you should see I've made the "warmup" phase an infinite loop. The idea is to make a RPC that sets warmUp to true, allowing for the infinite loop to terminate. The test can only finish if the RPC is allowed and completed while the recover is on-going, just what we wanted to prove. So, my question is: How can I invoke a RPC from that test?

Almost there!

**Comment by dmvk** [comment]
Hi @edu05, I've tried to outline the acceptance test in the [following gist](https://gist.github.com/dmvk/b707af75533f7046ccd1c633f3576476). It's just a quick draft to get you rough idea how to approach the issue, feel free to reuse any of the code.

**Comment by edu05** [comment]
Hi @dmvk while writing the acceptance test I found a couple of things that don't quite make sense to me at the moment. Could you help please?

1. I have found the new call to the recover method to still be in the JobMaster's main thread, not outside of it as desired. You can see this by debugging the new IT I added to the PR with a breakpoint inside recover. I'm attaching a sample image, notice how the call to recover is made from SchedulerUtils (as intended) but that call is in turn made from inside JobMaster's main thread, not outside.
debug

2. Even if the call was made from a separate thread, the first call to recover would only "warm up" for the period of time before the second call to recover via CheckpointCoordinator. If the delay between both calls is shorter than the time it takes for the first recover to execute, the JobMaster will become stalled at that point and unable to take RPC calls.

Does that make sense?

**Comment by dmvk** [comment]
Hi @edu05,

From the attached screenshot you can see that JobMaster initialization happens in "jobmanager-future-thread-1", before it confirms leadership. This is not a concern, because until leadership is confirmed, we just fail any RPC with ...

```
org.apache.flink.runtime.dispatcher.UnavailableDispatcherOperationException: Unable to get JobMasterGateway for initializing job. The requested operation is not available while the JobManager is initializing.
```

The actual problem is, when there is an actual recovery, that happens over RPC (eg. task failure). Every JobMaster is tied to a single RPC thread (this is what is referred to as a mainThread in JobMaster's code, which may be a little confusing) and if this thread is blocked, is can not response to any other RPC.

> Even if the call was made from a separate thread, the first call to recover would only "warm up" for the period of time before the second call to recover via CheckpointCoordinator. If the delay between both calls is shorter than the time it takes for the first recover to execute, the JobMaster will become stalled at that point and unable to take RPC calls.

Call to `recover()` is blocking, so JobMaster initialization won't happen until this initial "warmup" call to `recover()` finishes. We confirm leadership after this. See `JobMasterServiceLeadershipRunner#createNewJobMasterServiceProcess` for more details.


---

The new `CheckpointStoreRPCITCase` you've introduced does not stress the "RPC" recovery code path, unlike the example I've send you. Also it would be nice to make the new test part of already existing `CheckpointStoreITCase` instead of introducing a new test class.

---

Since the 1.14 release is getting closer, and there is still lot of tasks to be done, I'd timebox this effort until Monday 26th (inclusive). If we're not able to make this work until then, I'd take this over, so we can move on to the next task.

This task is really complex and requires lot of context, so I hope you won't have any hard feelings if we do this.

---

Good job so far! ;)

**Comment by edu05** [comment]
Hi @dmvk 

> This task is really complex and requires lot of context, so I hope you won't have any hard feelings if we do this.

For the more complex tasks is there a way the open source community can collaborate more closely with you? I feel like some diagrams would have been useful. Building in house/private software is hard enough to require closely knit collaboration between team members, could  we replicate that approach in open source?


> Since the 1.14 release is getting closer, and there is still lot of tasks to be done, I'd timebox this effort until Monday 26th (inclusive). If we're not able to make this work until then, I'd take this over, so we can move on to the next task.

There doesn't seem to be much choice, no hard feelings.


I've pushed the latest progress, mainly going back to your suggestion of putting both test cases into the same file where the RPC flow is executed and changing `started` from a `boolean` to an `int`. This is only because the `recover()` method is now called twice before the `FailingMapper` code is even executed.

At the moment the test gets stuck in `jobClient.getJobStatus()`, the reason for this is the thread invoking the RPC is blocked at the `recover()` stage, so it never gets to even try to fulfill the call. I get the idea is precisely to prove the RPC thread is free to make calls but that isn't possible if it's the one that is stuck from the action of invoking `finishedRecovering.await()`. Images attached.

bug2
The thread getting blocked is `flink-akka.actor.default-dispatcher-8` and it is that same thread that is expected to be invoking the RPC.

debug3

I tried switching a few things to no avail. If the idea is for the new call to `recover()` to solve said problem then I'd expect `finishedRecovering.await()` to happen on the thread that's invoking the new call to `recover()`. But that only happens once at the beginning of the test and can not get stuck otherwise the test just loops indefinitely at that point. Should it get called a second time during recovery?


> Good job so far! ;)

Thanks, let's see if we can get this one in the bag on Monday, or today if you're available :)

**Comment by dmvk** [comment]
> For the more complex tasks is there a way the open source community can collaborate more closely with you? I feel like some diagrams would have been useful. Building in house/private software is hard enough to require closely knit collaboration between team members, could we replicate that approach in open source?

I think subscribing to `dev@` mailing list is the best place to get you kick started.

All the high-level design decision go trough a formal, well documented process where everyone can participate. [Here](https://cwiki.apache.org/confluence/display/FLINK/Flink+Improvement+Proposals) is complete list of the design documents (FLIPS), which is probably the best source of the documentation for Flink internals so far.

Also, it's usually possible to find some context in JIRA issues (`git blame` helps here as we try to have each commit linked with a JIRA issue).

To sum up dev@apache.flink.org + confluence + JIRA + github, are the preferred channels for communication (we try to stick to an asynchronous communication as much as possible, because the developer time and focus is the most precious asset we have).

**Comment by tillrohrmann** [comment]
Merged as part of #16484.
